### PR TITLE
feat: add experiments.layer

### LINKED ISSUE #4398: Document webpack change: add experiments.layers
*A pull request by @sokra was merged and maintainers requested a documentation change.*

See pull request: https://github.com/webpack/webpack/pull/12327

---

* add experiments.layers
* add layers to define different module layers
* support module layers in stats
* add layer support for NormalModules

fixes #11466

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
* `experiments.layers: true` enables the experimental feature
* `entry.xxx.layer: "name-of-layer"` allows to set the layer for an entrypoint
* `module.rules[].issuerLayer: RuleCondition` allows to filter/match by layer of issuer
* `module.rules[].layer: "name-of-layer"` allows to set the layer of an module
* `splitChunks.cacheGroups.xxx.layer` allows to select modules by layer
* `externals.byLayer["name-of-layer"]` allows to specify externals by layer
* `externals` function has an additional argument property `contextInfo` which contains info about the issuer (e. g. the layer)

### PR DESCRIPTION
**Pull request by anshumanv**
Fix #4398 

- [x] Read and sign the [CLA][1]. PRs that haven't signed it won't be accepted.
- [x] Make sure your PR complies with the [writer's guide][2].
- [x] Review the diff carefully as sometimes this can reveal issues.
- [x] Do not abandon your Pull Request: [Stale Pull Requests][3].
- __Remove these instructions from your PR as they are for your eyes only.__

[1]: https://cla.js.foundation/webpack/webpack.js.org
[2]: https://webpack.js.org/contribute/writers-guide/
[3]: https://webpack.js.org/contribute/#pull-requests

### CODE REVIEW

**Comment by chenxsan** [comment]
@anshumanv Can you take care of the CI?

**Comment by anshumanv** [comment]
@chenxsan fixed!

File: src/content/configuration/module.mdx [hunk 1]
```diff
@@ -193,6 +193,45 @@ import A from './a.js';
...
+      }
+    ]
+  }
+};
+```
+
+
+## `Rule.layer`
+
+Name of layer.
```
Comment (chenxsan) [inline src/content/configuration/module.mdx:218]:
```suggestion
Specify the layer in which the module should be placed in.
```

File: src/content/configuration/entry-context.md [hunk 2]
```diff
@@ -82,6 +82,7 @@ module.exports = {
       filename: 'pages/personal.js',
       dependOn: 'shared',
       chunkLoading: 'jsonp',
+      layer: 'name of layer',
```
Comment (chenxsan) [inline src/content/configuration/entry-context.md:85]:
```suggestion
      layer: 'name of layer', // set the layer for an entry point
```

File: src/content/plugins/split-chunks-plugin.md [hunk 3]
```diff
@@ -188,6 +189,15 @@ Size threshold at which splitting is enforced and other restrictions (minRemaini
...
 W> `splitChunks.minRemainingSize` only takes effect when a single chunk is remaining.
 
+
+### `splitChunks.layer`
+
+#### `splitChunks.cacheGroups.{cacheGroup}.layer`
+
+`RegExp` `string` `Function`
+
+Allows to select modules by layer.
```
Comment (chenxsan) [inline src/content/plugins/split-chunks-plugin.md:199]:
```suggestion
Assign modules to a cache group by module layer.
```

File: src/content/configuration/externals.md [hunk 4]
```diff
@@ -144,6 +145,21 @@ module.exports = {
...
+
+`string` `RegExp`
+
+W> Specify externals depending on the layer.
+
+```javascript
+module.exports = {
+  externals : {
+    byLayer : {
+      root: 'test-layer',
```
Comment (chenxsan) [inline src/content/configuration/externals.md:158]: See example here https://github.com/webpack/webpack/blob/master/test/configCases/layer/rules/webpack.config.js#L54

File: src/content/configuration/module.mdx [hunk 5]
```diff
@@ -193,6 +193,45 @@ import A from './a.js';
...
+
+__webpack.config.js__
+
+```js
+module.exports = {
+  // ...
+  module: {
+    rules: [
+      {
+        issuerLayer: RuleCondition,
```
Comment (chenxsan) [inline src/content/configuration/module.mdx:208]: See example https://github.com/webpack/webpack/blob/master/test/configCases/layer/rules/webpack.config.js#L24

File: src/content/configuration/module.mdx [hunk 6]
```diff
@@ -193,6 +193,45 @@ import A from './a.js';
...
+
+__webpack.config.js__
+
+```js
+module.exports = {
+  // ...
+  module: {
+    rules: [
+      {
+        layer: 'test-layer',
```
Comment (chenxsan) [inline src/content/configuration/module.mdx:228]: See example https://github.com/webpack/webpack/blob/master/test/configCases/layer/rules/webpack.config.js#L16

File: src/content/configuration/externals.md [hunk 7]
```diff
@@ -144,6 +145,21 @@ module.exports = {
 
 This syntax is used to describe all the possible ways that an external library can be made available. `lodash` here is available as `lodash` under AMD and CommonJS module systems but available as `_` in a global variable form. `subtract` here is available via the property `subtract` under the global `math` object (e.g. `window['math']['subtract']`).
 
+### `byLayer`
+
+`string` `RegExp`
```
[the remark this answers is not in the document]
Comment (anshumanv) [reply src/content/configuration/externals.md:150]: fixed

File: src/content/configuration/module.mdx (hunk 1 above)
Comment (anshumanv) [reply src/content/configuration/module.mdx:218]: Done

File: src/content/configuration/module.mdx [hunk 8]
```diff
@@ -193,6 +193,46 @@ import A from './a.js';
...
+        issuerLayer: 'other-layer',
+      }
+    ]
+  }
+};
+```
+
+
+## `Rule.layer`
+
```
Comment (chenxsan) [inline src/content/configuration/module.mdx:217]:
```suggestion

`string`

```

File: src/content/configuration/module.mdx [hunk 9]
```diff
@@ -193,6 +193,46 @@ import A from './a.js';
 This option can be used to apply loaders to the dependencies of a specific module or set of modules.
 
 
+## `Rule.issuerLayer`
+
+Allows to filter/match by layer of issuer.
```
Comment (chenxsan) [inline src/content/configuration/module.mdx:198]:
```suggestion
Allows to filter/match by layer of the issuer.
```

File: src/content/configuration/externals.md [hunk 10]
```diff
@@ -144,6 +145,23 @@ module.exports = {
 
 This syntax is used to describe all the possible ways that an external library can be made available. `lodash` here is available as `lodash` under AMD and CommonJS module systems but available as `_` in a global variable form. `subtract` here is available via the property `subtract` under the global `math` object (e.g. `window['math']['subtract']`).
 
+### `byLayer`
```
Comment (chenxsan) [inline src/content/configuration/externals.md:148]: Let's move this section to the end. All others are talking about types of `externals`, it's sort of abrupt to insert `byLayer` here.

Comment (anshumanv) [reply src/content/configuration/externals.md]: fixed

**Comment by anshumanv** [comment]
fixed all

**Review by chenxsan — APPROVED** [review]
Looks good to me now, thanks.

**Pull request merged**